src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -716,31 +728,27 @@ fn ranges_subtract_pattern<'a, 'tcx>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
         let mut ranges: Vec<_> = ranges.into_iter().filter_map(|r| {
             Interval::from_ctor(&r).map(|i| i.into_inner())
         }).collect();
+        let ty = pat_interval.ty;
+        let (pat_interval_lo, pat_interval_hi) = pat_interval.into_inner();
         for (subrange_lo, subrange_hi) in ranges {
-            if pat_interval.lo > subrange_hi || pat_interval.hi < subrange_lo {
+            if pat_interval_lo > subrange_hi || subrange_lo > pat_interval_hi  {
                 // The pattern doesn't intersect with the subrange at all,
                 // so the subrange remains untouched.
                 remaining_ranges.push((subrange_lo, subrange_hi));
-            } else if pat_interval.lo <= subrange_lo && pat_interval.hi >= subrange_hi {
-                // The pattern entirely covers the subrange of values,
-                // so we no longer have to consider this subrange_
-            } else if pat_interval.lo <= subrange_lo {
-                // The pattern intersects a lower section of the subrange,
-                // so only the upper section will remain.
-                remaining_ranges.push((pat_interval.hi + 1, subrange_hi));
-            } else if pat_interval.hi >= subrange_hi {
-                // The pattern intersects an upper section of the subrange,
-                // so only the lower section will remain.
-                remaining_ranges.push((subrange_lo, pat_interval.lo - 1));
             } else {
-                // The pattern intersects the middle of the subrange,
-                // so we create two ranges either side of the intersection.)
-                remaining_ranges.push((subrange_lo, pat_interval.lo));
-                remaining_ranges.push((pat_interval.hi, subrange_hi));
+                if pat_interval_lo > subrange_lo {
+                    // The pattern intersects an upper section of the
+                    // subrange, so a lower section will remain.
+                    remaining_ranges.push((subrange_lo, pat_interval_lo - 1));
+                }
+                if pat_interval_hi < subrange_hi {
+                    // The pattern intersects a lower section of the
+                    // subrange, so an upper section will remain.
+                    remaining_ranges.push((pat_interval_hi + 1, subrange_hi));
```

Comment:
`remaining_ranges` could be a `Vec<RangeInclusive<u128>>`, i.e. use `..=` syntax.